environs/manual/sshprovisioner/sshprovisioner.go

```diff
@@ -210,29 +269,41 @@ grep MemTotal /proc/meminfo
 cat /proc/cpuinfo`
 
 // gatherMachineParams collects all the information we know about the machine
-// we are about to provision. It will SSH into that machine as the ubuntu user.
+// we are about to provision. It will SSH using the provisioning user.
 // The hostname supplied should not include a username.
 // If we can, we will reverse lookup the hostname by its IP address, and use
 // the DNS resolved name, rather than the name that was supplied
-func gatherMachineParams(hostname string) (*params.AddMachineParams, error) {
-
+func gatherMachineParams(hostname string, provisioningUserName string, privateKey string) (*params.AddMachineParams, error) {
 	// Generate a unique nonce for the machine.
 	uuid, err := utils.NewUUID()
 	if err != nil {
 		return nil, err
 	}
 
-	provisioned, err := checkProvisioned(hostname)
-	if err != nil {
-		return nil, errors.Annotatef(err, "error checking if provisioned")
+	var provisioned bool
+	var provisionCheckErr error
+	if provisioningUserName != "" || privateKey != "" {
+		provisioned, provisionCheckErr = checkProvisionedAsUser(hostname, provisioningUserName, privateKey)
+	} else {
+		provisioned, provisionCheckErr = checkProvisioned(hostname)
+	}
```

Comment:
Why this?
If you just call `checkProvisionedAsUser(hostname, provisioningUserName, privateKey)` and the user name or key is empty, it does the right thing. No need to if then else here.